tests/util/caches/test_responsecache.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2021 Vector Creations Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from twisted.internet.task import Clock
+
+from synapse.util import Clock as SynapseClock
+from synapse.util.caches.response_cache import ResponseCache
+
+from tests.unittest import TestCase
+
+# few notes about test naming here:
+# 'wait': denotes tests that have an element of "waiting" before its wrapped result becomes available
+# 'expire': denotes tests that test expiry after assured existence
+
+
+class DeferredCacheTestCase(TestCase):
+    def setUp(self):
+        self.reactor = Clock()
+        self.synapse_clock = SynapseClock(self.reactor)
```

Comment:
I think you want to use `tests.server.get_clock` here (and it would be nice to name them the same as `reactor` and `clock` (to match `HomeserverTestCase`).

It might even make sense to create the cache object here and then use it for each test (this simplifies each individual test-case, which makes it easier to see what each test-case is testing). I know the timeouts are currently different on them, but I think the tests could be slightly modified to take that into account.

Reply:
I'll use `get_clock`, but i wont move the cache here, as they need to be variable for tests (see the naming and `timeout_ms` on each. :+1: 

One thing I *will* do is make it a separate function (with accepts a name and optionally `ms=`), so that future tests can just use that instead of instantiating it directly every time.

Reply:
I changed the instantiation to `with_cache` to make it easier on the eyes, please notify if this is wrong.